### DISCUSSION TITLE
Using the library with any open source llm

### DISCUSSION THREAD

**Discussion by harsh3986** [sinaptik-ai/pandas-ai#373] [Q&A]
How can I use pandas ai with any open source llms or let's say any model from hugging face ?

**Comment by sandiemann** [comment]
this doc might be out to date but you can initiate the HF llm models Starcoder or Falcon as mentioned [here](https://pandas-ai.readthedocs.io/en/latest/llms/#huggingface-models)

**Reply by harsh3986** [reply]
Thanks @sandiemann.  Suppose I have created a fine tuned starcoder model, how will I use it in that case ?

**Reply by sandiemann** [reply]
sample code:
```
from pandasai import PandasAI
from pandasai.llm.starcoder import Starcoder

df = pd.read_csv("examples/data/Loan payments data.csv")

#llm = OpenAI()
llm = Starcoder()
pandas_ai = PandasAI(llm, verbose=True, conversational=True)
response = pandas_ai(df, "How many loans are from men and have been paid off?")
print(response)
```

output logs:
```
2023-07-18 23:51:05 [INFO] Question: How many loans are from men and have been paid off?
2023-07-18 23:51:05 [INFO] Running PandasAI with starcoder LLM...
2023-07-18 23:51:05 [INFO] Prompt ID: 392db83b-6b36-4d71-98f7-1b5ab0bfa406
2023-07-18 23:51:07 [INFO] 
                        Code generated:
                        ```
                        df[df['Gender']=='male']
                        df[df['Gender']=='male']['loan_status'].value_counts()
                        ```
                    
2023-07-18 23:51:07 [INFO] 
Code running:

df[df['Gender'] == 'male']
df[df['Gender'] == 'male']['loan_status'].value_counts()

2023-07-18 23:51:07 [INFO] Answer: PAIDOFF               247
COLLECTION             90
COLLECTION_PAIDOFF     86
Name: loan_status, dtype: int64
```

**Comment by mspronesti** [comment]
@harsh3986  In the case of a fine-tuned LLM, PandasAI does not natively support any solution as it relies on the hosted inference api for hf models. However, since it is integrated with `langchain`, you might try to load your fine-tuned StarCoder as a `langchain` model (e.g. assuming you uploaded it to huggingface hub, you can use this [model](https://github.com/hwchase17/langchain/blob/master/langchain/llms/huggingface_hub.py)) and pass it to the `PandasAI` class, which internally converts it.
Hope this helps.
